docs/papers/utils/minipaper_cracking_symmetry.py

```
@@ -0,0 +1,353 @@
+    print(f"📦 Installing packages: {' '.join(required)}")
+    subprocess.run(
+        [sys.executable, "-m", "pip", "install"] + required,
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+    )
+    print("✅ All packages installed.")
+
+install_packages() # do it now for convenience
+
+# Group into contiguous non-NaN ranges (i.e., success ranges!)
+def group_success_ranges(theta1_vals, theta2_vals):
```

Ruff (0.11.9)

Missing return type annotation for public function `group_success_ranges`

(ANN201)